### PR TITLE
chore(flags): add billing to rust endpoint for local_evaluation

### PR DESCRIPTION
**Pull request by patricio-posthog**
# Summary

- Add billing quota enforcement to the Rust `/flags/definitions` endpoint, returning 402 when the team exceeds their feature flag request quota (matching Django's `/local_evaluation` behavior)
- Filter non-billable flags (survey and product tour targeting flags) from usage tracking, so only requests with actual feature flags count toward billing
- This is Task 5 of the `/local_evaluation` → Rust migration #49564 
- 
## Changes

**Quota limiting** (`flag_definitions.rs`):
- Check `FeatureFlagsLimiter.is_limited(token)` after rate limiting, before fetching cache
- Returns `FlagError::ClientFacing(ClientFacingError::BillingLimit)` → HTTP 402
- Uses the same `FeatureFlagsLimiter` already wired up for `/flags`

**Billable flag filtering** (`flag_definitions.rs`):
- Moved usage tracking (`increment_request_count`) to after cache fetch so we can inspect the response
- Added `has_billable_flags()` helper that checks if the cached response contains any flags whose key doesn't start with `survey-targeting-` or `product-tour-targeting-`
- Matches Django's filtering logic in `local_evaluation`

> **Note on billable flag filtering:** `/local_evaluation` returns ALL flag definitions in a single request, always counted as 1. The filtering only skips counting if a team has *exclusively* survey/product tour targeting flags and zero real feature flags, an edge case that's unlikely in practice. We match Django's behavior for parity.

**Tests**:
- 7 unit tests for `has_billable_flags`: regular flags, only surveys, only product tours, mixed, empty array, missing key, combined survey+tour
- 1 integration test: creates a billing-limited team via mock Redis, verifies `/flags/definitions` returns 402

## How Django `/local_evaluation` compares

| Aspect | Django | Rust (after this PR) |
|--------|--------|---------------------|
| Quota check | `list_limited_team_attributes` → 402 | `FeatureFlagsLimiter.is_limited` → 402 |
| Usage tracking | `increment_request_count` with `LOCAL_EVALUATION` type | Same function, same Redis keys |
| Billable filtering | Skips count if all flags are survey/tour | Same logic via `has_billable_flags()` |

## Test plan

- [x] Unit tests for `has_billable_flags` pass
- [x] Integration test for billing-limited 402 response passes
- [x] `cargo fmt`, `cargo clippy` clean

### CODE REVIEW

File: rust/feature-flags/src/api/flag_definitions.rs [hunk 1]
```diff
@@ -362,9 +378,81 @@ async fn authenticate_flag_definitions(
...
+    }
+
+    #[test]
+    fn test_has_billable_flags_only_survey_and_tour_flags() {
+        let response = json!({"flags": [
+            {"key": "survey-targeting-abc"},
+            {"key": "product-tour-targeting-xyz"},
+        ]});
+        assert!(!has_billable_flags(&response));
+    }
```
[the remark this answers is not in the document]
Comment (patricio-posthog) [reply rust/feature-flags/src/api/flag_definitions.rs]: I'll do this.

File: rust/feature-flags/src/api/flag_definitions.rs [hunk 2]
```diff
@@ -362,9 +378,81 @@ async fn authenticate_flag_definitions(
...
+/// matching Django's `local_evaluation` billing filter. The cached response
+/// has a `"flags"` array where each entry has a `"key"` field.
+fn has_billable_flags(response: &Value) -> bool {
+    let Some(flags) = response.get("flags").and_then(|f| f.as_array()) else {
+        return false;
+    };
+
+    flags.iter().any(|flag| {
+        let key = flag.get("key").and_then(|k| k.as_str()).unwrap_or("");
+        !key.starts_with(SURVEY_TARGETING_FLAG_PREFIX)
```
Comment (haacked) [inline rust/feature-flags/src/api/flag_definitions.rs]: `suggestion`: This reimplements the same "is this flag billable?" predicate from `handler/billing.rs::contains_billable_flags`. The input types differ (raw JSON vs typed structs), but the core rule is identical. If a third non-billable prefix gets added, one site will get missed.

Consider extracting just the predicate into a shared function:

```rust
// In flag_analytics.rs (next to the prefix constants)
pub fn is_billable_flag_key(key: &str) -> bool {
    !key.starts_with(SURVEY_TARGETING_FLAG_PREFIX)
        && !key.starts_with(PRODUCT_TOUR_TARGETING_FLAG_PREFIX)
}
```

Then both `has_billable_flags` here and `contains_billable_flags` in billing.rs call the same function. The JSON traversal stays here, the business rule lives in one place.

File: rust/feature-flags/tests/test_flag_definitions.rs [hunk 3]
```diff
@@ -1674,3 +1674,50 @@ async fn test_etag_graceful_degradation_without_stored_etag() {
...
+        .unwrap();
+
+    let status = response.status();
+    let body_text = response.text().await.unwrap();
+
+    assert_eq!(
+        status, 402,
+        "Should return 402 when billing quota is exceeded. Body: {body_text}"
+    );
+    assert_eq!(
```
Comment (haacked) [inline rust/feature-flags/tests/test_flag_definitions.rs]: `question`: Django returns a JSON body for 402: `{"type": "quota_limited", "detail": "...", "code": "payment_required"}`. This returns plain text. If any SDK parses the 402 body (e.g., checking for `"type": "quota_limited"`), it'll break when traffic shifts to this endpoint.

Worth confirming SDKs key on the status code, not the body. If that's confirmed, a quick comment here noting the intentional divergence would help future readers.

File: rust/feature-flags/src/api/flag_definitions.rs [hunk 4]
```diff
@@ -137,6 +133,26 @@ pub async fn flags_definitions(
     // Retrieve cached response from HyperCache (always with cohorts)
     let cached_response = get_from_cache(&state, &team_key, team.id).await?;
 
+    // Record usage for billing, filtering out non-billable flags (surveys, product tours)
```
Comment (haacked) [inline rust/feature-flags/src/api/flag_definitions.rs]: `nit`: Worth a brief comment here noting that the placement after the cache fetch (and after the ETag/304 path) is intentional. 304 responses skip billing, which matches Django's `/local_evaluation` behavior. The old Rust code counted 304s, so this is a deliberate fix.

**Review by haacked — COMMENTED** [review]
Nice work! I'll approve after you address the inline question about the 402 response body.

File: rust/feature-flags/src/api/flag_definitions.rs (hunk 2 above)
Comment (patricio-posthog) [reply rust/feature-flags/src/api/flag_definitions.rs]: Ah yes, I didnt know this. Done!

File: rust/feature-flags/src/api/flag_definitions.rs (hunk 4 above)
Comment (patricio-posthog) [reply rust/feature-flags/src/api/flag_definitions.rs]: Added

File: rust/feature-flags/tests/test_flag_definitions.rs (hunk 3 above)
Comment (patricio-posthog) [reply rust/feature-flags/tests/test_flag_definitions.rs]: Good catch. I checked andthe SDK only parses the status, not the body. However, just to be sure, I updated the body to be json instead of only a string. We can change it to string in the future if we want to simplify (though I don't see a good reason to do it)

**Review by haacked — APPROVED** [review]
Excellent!
